dbms/src/Storages/Page/V3/PageDirectory.cpp

```diff
@@ -117,6 +119,88 @@ void VersionedPageEntries::createNewEntry(const PageVersion & ver, const PageEnt
         ErrorCodes::PS_DIR_APPLY_INVALID_STATUS);
 }
 
+PageIdV3Internal VersionedPageEntries::createUpsertEntry(const PageVersion & ver, const PageEntryV3 & entry)
+{
+    auto page_lock = acquireLock();
+
+    if (type == EditRecordType::VAR_DELETE)
+    {
+        type = EditRecordType::VAR_ENTRY;
```

Comment:
I don't understand why change `VAR_DELETE` to `VAR_ENTRY`. Can you add some comments?

Reply:
`VAR_DELETE` is the init state when creating a new version list, so here just turn `VAR_DELETE` into `VAR_ENTRY` and create a new <`ver, entry>`.

But no matter tiflash calls this function during restart or running, we should ensure that there exist an "old-entry" before this upsert entry. So `VAR_DELETE` is not a valid state. I will remove these lines and let it throw exception.